main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala

```diff
@@ -3004,21 +3004,27 @@ object Weeder2 {
 
     private def visitConstraint(tree: Tree)(implicit sctx: SharedContext): Validation[TraitConstraint, CompilationMessage] = {
       expect(tree, TreeKind.Type.Constraint)
-      flatMapN(pickQName(tree), Types.pickType(tree)) {
+      mapN(pickQName(tree), Types.pickType(tree)) {
         (qname, tpe) =>
           // Check for illegal type constraint parameter
-          if (!isAllVariables(tpe)) {
-            Validation.toHardFailure(IllegalTraitConstraintParameter(tree.loc))
-          } else {
-            Validation.success(TraitConstraint(qname, tpe, tree.loc))
-          }
+          val (tpe1, errors) = replaceWithTypeVars(tpe)
+          errors.headOption.map(loc => sctx.errors.add(IllegalTraitConstraintParameter(loc)))
+          TraitConstraint(qname, tpe1, tree.loc)
       }
     }
 
-    private def isAllVariables(tpe: Type): Boolean = tpe match {
-      case _: Type.Var => true
-      case Type.Apply(t1, ts, _) => isAllVariables(t1) && isAllVariables(ts)
-      case _ => false
+    private def replaceWithTypeVars(tpe: Type): (Type, List[SourceLocation]) = {
```

Comment:
should probably rename this function to something more descriptive and add doc

Reply:
Yes please. 

Also explain what this whole thing does. I am unfamiliar with both the old and new code.

Reply:
Was adding description as you wrote :))

Reply:
https://github.com/flix/flix/pull/9082#issue-2629641682